keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1636 @@
+node is found.
+
+The hardware that is expected to need this more flexible allocation
+approach is going to be used by only a small subset of the pods in the
+cluster and those pods are likely to run for extended periods of time,
+so this is not a major concern.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Implementation
+
+![components](./components.png)
+
+Several components must be implemented or modified in Kubernetes:
+- The new API must be added to kube-apiserver.
+- A new controller in kube-controller-manager which generates
+  ResourceClaims from Pod ResourceClaimTemplates, similar to
+  https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/volume/ephemeral
+- A kube-scheduler plugin must detect Pods which reference a
+  ResourceClaim (directly or through a template) and ensure that the
+  resource is allocated before the Pod gets scheduled, similar to
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/scheduling/scheduler_binder.go
+- Kubelet must be extended to retrieve information from ResourceClaims
+  and to invoke a local resource driver method. That method returns CDI device ID(s)
+  which then must be passed to the container runtime.
+
+For a resource driver the following components are needed:
+- *Resource driver controller*: a central component which handles resource allocation
+  by watching ResourceClaims and updating their status once it is done with
+  allocation.
+- *Resource kubelet plugin*: a component which cooperates with kubelet to prepare
+  the usage of the resource on a node.
+
+The utility library will be developed outside of Kubernetes and does not have
+to be used by drivers, therefore it is not described further in this KEP.
+
+### Resource allocation flow
+
+The following diagram shows how resource allocation works for a resource that
+gets defined inline in a Pod. For a full definition of ResourceClass,
+ResourceClaim and ResourceClaimTemplate see the [API](#API) section below.
+
+Several of these operations may fail without changing the system state. They
+then must be retried until they succeed or something else changes in the
+system, like for example deleting objects. These additional state transitions
+are not shown for the sake of simplicity.
+
+![allocation](./allocation.png)
```

Comment:
I don't quite follow how "immediate" allocation works. For the "delayed" case, we have to pick a node for the pod before the resource gets allocated, but (as far as I can see from this diagram) this doesn't seem to be the case for "immediate" allocation. Can you clarify this for me?

More specifically, I could see how this would work for ResourceClaims created outside of a PodSpec, but I don't see how it would work (or what advantage it would provide) for ResourceClaims created in response to a pod being created.

Reply:
There's no advantage of using "immediate" for ResourceClaims created from an inline template in a Pod. For that use case, "delayed" is what a users should be using.

"Immediate" is relevant when creating a ResourceClaim separately from a Pod as a stand-alone object with its own lifecycle.